### ISSUE TITLE
Proposal: Function multi-versioning

### ISSUE THREAD

**Issue by bheads** [zig-lang/zig#1018]
A really interesting concept is function multi-versioning. The general idea is to support implementing multiple versions of a function for different hardware and having the correct version of the function selected at run time. Made up sample code:

```C
pub  fn someMathFunction(vec: Vector) Vector [target: sse4.2]
{
   // optimized for SSE 4.2
}

pub  fn someMathFunction(vec: Vector) Vector [target: avx2]
{
   // optimized for avx2 
}

pub  fn someMathFunction(vec: Vector) Vector [target: default]
{
   // no asm/intrinsics optimization
}

// later on

const v = giveMeAVect();
const v2 = someMathFunction(v); // calls the best version based on run time selection
```

There are ways to simulate this using function pointers, but the compiler would be better at optimizing this, plus implementing that over and over by hand would suck.  

LLVM https://sourceware.org/glibc/wiki/GNU_IFUNC 
GCC https://lwn.net/Articles/691932/

**Comment by andrewrk** [comment]
This use case is in scope of Zig. Whether we use LLVM IFuncs or something else is still up for research, as well as the syntax and how it fits in with comptime and other features.

**Comment by bnoordhuis** [comment]
Ifuncs only work on ELF platforms (and not all of them either) so Zig would need a fallback for platforms where they're not supported, such as Windows, FreeBSD < 12, etc.

I'd attack this bottom-up: start with a generic solution and later on add ifuncs as a compiler optimization.

**Comment by PavelVozenilek** [comment]
1. In the build system allow targets with different properties (SSE/cache size/main memory size/GPU/...).
2. Allow specialized constants/data structures/functions by these properties.
3. Generate "fat executable" where all targets are generated into one file, and the most fitting target gets selected at the startup.
4. For very specialized cases allow to offer own selection mechanism (by running a benchmark at the startup). 

Using hidden function pointers set by the compiler for individual functions would hog cache friendliness and destroy potential for inlining.

**Comment by binary132** [comment]
The way this is solved in Go is with a [build tag](https://golang.org/pkg/go/build/#hdr-Build_Constraints) at the file level.  When the build runs, some build properties are defined depending on the platform and optional build arguments.  The entire file is either included or excluded depending on its build tags.  This makes for a very simple approach, and it also forces the user to put all platform-specific stuff in a single file.

**Comment by tiehuis** [comment]
@binary132 That is different to the issue here since it is a compile-time selection. This is possible in Zig right now by using standard if statements with comptime values. This issue is about function selection at runtime, based on runtime cpu feature detection.

This is useful to generate a single portable binary that runs on different cpu processor architectures (e.g. core 2 duo vs. i7 skylake), while still allowing cpu-specific performance optimizations for hot functions.

**Comment by 0joshuaolson1** [comment]
> function selection at runtime, based on runtime cpu feature detection

Or when using Vulkan? :smirk:

**Comment by bronze1man** [comment]
I proposal follow way to solve this stuff to make the code more easy to read and write:
```
pub  fn someMathFunction(vec: Vector) Vector
{
    @useLLVMIFuncs
    if (builtin.hasSse42){
      ...
    }else if (builtin.hasAvx2){
      ...
    }else{
      ...
    }
}
```
And the compiler can generate 3 versions of this function and generate the LLVM IFuncs base on that information.

Solve multi-versioning problem like golang make the code difficult to read and write and refactor. (Do I define all the need versions I need? Do I change all the function name of all the version?)

**Comment by tiehuis** [comment]
I think the switch as specified by @bronze1man seems most in line with how zig would work, given it uses all standard syntax. Ignoring the ifunc optimization, first, we need cpu feature detection of sorts. I've implemented below some example code which provides runtime cpu feature detection to get a better idea of potential issues.

### Example

```
// initialized globally somewhere, could do locally but minor cost
var cpu = CpuFeatures.init();

pub fn memchr(buf: *const u8, c: u8) usize {
    if (cpu.hasFeature(FeatureX86.avx512)) {
        return memchr_avx512(buf, c);
    } else if (cpu.hasFeature(FeatureX86.sse2)) {
        return memchr_sse2(buf, c);
    } else {
        return memchr_generic(buf, c);
    }
}
```

### Implementation

**cpuid.c**

```c
// This is done separately for now since zig's multi-return inline asm was a pain.
#include <cpuid.h>

int cpuid(unsigned int leaf, unsigned int *eax, unsigned int *ebx, unsigned int *ecx, unsigned int *edx)
{
    return __get_cpuid(leaf, eax, ebx, ecx, edx);
}
```

**cpu.zig**

```
const std = @import("std");
const builtin = @import("builtin");

// Runtime cpu feature detection.
//
// This is currently implemented for x86/x64 targets. For generic targets, the features
// returned will be compile-time false and will not use any code space.

comptime {
    std.debug.assert(@memberCount(FeatureX86) == 224);
}

// See https://en.wikipedia.org/wiki/CPUID
pub const FeatureX86 = enum {
    // eax = 1, output: edx
    fpu,
    vme,
    de,
    pse,
    tsc,
    msr,
    pae,
    mce,
    cx8,
    apic,
    _reserved1,
    sep,
    mtrr,
    pge,
    mca,
    cmov,
    pat,
    pse_36,
    psn,
    clfsh,
    _reserved2,
    ds,
    acpi,
    mmx,
    fxsr,
    sse,
    sse2,
    ss,
    htt,
    tm,
    ia64,
    pbe,

    // omitted remaining features for brevity
    ....
};

// Implemented in C until multi-output asm is easier. See: #215.
extern fn cpuid(leaf: c_uint, eax: *c_uint, ebx: *c_uint, ecx: *c_uint, edx: *c_uint) c_int;

pub const CpuFeatures = struct {
    buf: [7]u32,

    pub fn init() CpuFeatures {
        var self = CpuFeatures{ .buf = []u32{0} ** 7 };

        switch (builtin.arch) {
            builtin.Arch.i386, builtin.Arch.x86_64 => {
                var eax: c_uint = undefined;
                var ebx: c_uint = undefined;
                var ecx: c_uint = undefined;
                var edx: c_uint = undefined;

                // We don't strictly need to check this since __get_cpuid does but our
                // implementation may not.
                const max_basic_cpu_leaf = cpuid(0, &eax, &ebx, &ecx, &edx);

                if (max_basic_cpu_leaf >= 1) {
                    if (cpuid(1, &eax, &ebx, &ecx, &edx) == 1) {
                        self.buf[0] = edx;
                        self.buf[1] = ecx;
                    }
                }

                if (max_basic_cpu_leaf >= 7) {
                    if (cpuid(7, &eax, &ebx, &ecx, &edx) == 1) {
                        self.buf[2] = ebx;
                        self.buf[3] = ecx;
                        self.buf[4] = edx;
                    }
                }

                const max_ext_cpu_leaf = cpuid(0x80000000, &eax, &ebx, &ecx, &edx);

                if (max_ext_cpu_leaf >= 1) {
                    if (cpuid(0x80000001, &eax, &ebx, &ecx, &edx) == 1) {
                        self.buf[5] = edx;
                        self.buf[6] = ecx;
                    }
                }
            },
            else => {
                // other targets return false for `hasFeature`
            },
        }

        return self;
    }

    // This would actually take a var, which would accept any platform Feature, i.e.
    // FeatureX86 or FeatureARM (is that applicable?).
    //
    // If compiling for a target without this feature we know at comptime that we can never
    // execute that feature and no arch-specific code is included.
    fn hasFeature(self: *const CpuFeatures, feature: FeatureX86) bool {
        // We require #868 for allowing runtime-selected functions to be allowed to run at
        // compile-time. Assuming something like D and its __ctfe for the moment.
        if (__ctfe) {
            return false;
        }

        switch (builtin.arch) {
            builtin.Arch.i386, builtin.Arch.x86_64 => {
                const n = @enumToInt(feature);
                return (self.buf[n >> 5] & (u32(1) << @intCast(u5, n & 0x1f))) != 0;
            },
            else => {
                return false;
            },
        }
    }
};
```

### Notes

Non-compatible architectures where the feature does not work are compile-time known which allows us to avoid compiling in incompatible branches.

For the following example:

```
fn runtimeSelectableFunction() void {
    if (cpu.hasFeature(FeatureX86.sse42)) {
        @compileLog("not allowed to compile to target which may have sse");
    } else {
        std.debug.warn("no sse\n");
    }
}

var cpu: CpuFeatures = undefined;

pub fn main() void {
    cpu = CpuFeatures.init();
    runtimeSelectableFunction();
}
```

```
$ zig build-exe cpu.zig
/home/me/src/cpuid/cpu.zig:356:9: error: not allowed to compile to target which may have sse
        @compileError("not allowed to compile to target which may have sse");

$ zig build-exe cpu.zig --target-arch armv7
# all ok
```

We require #868 to allow for suitable fallback implementations where not otherwise specified. Unfortunately I don't think this can be avoided in this case.

### Links
 - #215 
 - #868

**Comment by emekoi** [comment]
does llvm have any cpu feature capability detection features we could leverage instead of rolling our own?

**Comment by shawnl** [comment]
The syntax is not OK. This feature must be built-in to the compiler because LLVM has to target the specified architecture for codegen and for good reasons these settings can only be done on a per-function basis. Also something like IFUNC can only be done per-function, as only functions have a clearly specified ABI/API (the calling convention). The comptime syntax can only be supported if no variables are allowed outside the `if` and this complication would make the language ugly. It needs to be a function attribute. Anyways, sorry for having such a strong opinion, but the syntax needs to be a calling convention declaration (that could also influence `-march`).

**Comment by shawnl** [comment]
> does llvm have any cpu feature capability detection features we could leverage instead of rolling our own?

yes, but it is x86-specific and we already have it in the zig tree as `./c_headers/cpuid.h`. For other architectures it may be (Linux-specific) `/proc/cpuinfo`. IFUNC has yet to be supported on non-x86 architectures.

**Comment by shawnl** [comment]
> We require #868 to allow for suitable fallback implementations where not otherwise specified. Unfortunately I don't think this can be avoided in this case.

We should probably model this after Linux's `AT_HWCAP` and `AT_HWCAP2`, which is not needed on the CSICy x86, but is needed on most other arches.

**Comment by daurnimator** [comment]
See also the GCC `target_clones` function attribute: https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-target_005fclones-function-attribute

> The target_clones attribute is used to specify that a function be cloned into multiple versions compiled with different target options than specified on the command line. The supported options and restrictions are the same as for target attribute.
> 
> For instance, on an x86, you could compile a function with target_clones("sse4.1,avx"). GCC creates two function clones, one compiled with -msse4.1 and another with -mavx.
> 
> On a PowerPC, you can compile a function with target_clones("cpu=power9,default"). GCC will create two function clones, one compiled with -mcpu=power9 and another with the default options. GCC must be configured to use GLIBC 2.23 or newer in order to use the target_clones attribute.
> 
> It also creates a resolver function (see the ifunc attribute above) that dynamically selects a clone suitable for current architecture. The resolver is created only if there is a usage of a function with target_clones attribute.
> 
> Note that any subsequent call of a function without target_clone from a target_clone caller will not lead to copying (target clone) of the called function. If you want to enforce such behaviour, we recommend declaring the calling function with the flatten attribute?

**Comment by zxubian** [comment]
I think there should be a way to disable this at compile-time.

The use cases discussed so far assume you don't know the specifics of your target CPU. However, there are cases when you write code knowing exactly what hardware it will run on.

If the standard library starts using function multi-versioning, yet you know at compile time that you will only need one specific version, you should be able to select that version at compile time and make sure that no others make it to the binary.

Otherwise, you'll waste executable size (and potentially cache penalties if this is implemented via pointers).

**Comment by Jarred-Sumner** [comment]
This has been a big pain point for Bun.

- https://github.com/oven-sh/bun/issues/1088
- https://github.com/oven-sh/bun/issues/1447
- https://github.com/oven-sh/bun/issues/1167
- https://github.com/oven-sh/bun/issues/684
- https://github.com/oven-sh/bun/issues/461
- https://github.com/oven-sh/bun/issues/197

Our current solution is to compile two versions of bun for Linux x64 and macOS x64. One which targets `haswell` and one which targets `westmere`.  We use an install script that detects if the computer supports AVX2 and chooses the right binary. We recompile all dependencies with the same CPU target set.

This unfortunately breaks when someone installs Bun without using the official install script that checks for this - like if using a package manager for a Linux distribution. Having to separate by CPU features makes it more difficult for people to distribute Bun.

**Comment by mitchellh** [comment]
@Jarred-Sumner I'm interested in this issue too but you can do runtime simd detection to avoid having to build two binaries. Its a bit of work but it isn't too bad. You have to abandon all of the Zig intrinsics and write raw assembly but I've been doing it some for some SIMD work and its been going along fine. Ping me on discord if you need help I have some example code I can share with you.

**Comment by daurnimator** [comment]
> Edit: here is how simdjson does ISA detection: https://sourcegraph.com/github.com/simdjson/simdjson/-/blob/include/simdjson/internal/isadetection.h

FYI zig already has cpuid code in the standard library https://github.com/ziglang/zig/blob/694d8831c31fca5b669b52d3f9f95a8792daee41/lib/std/zig/system/x86.zig#L25 
Though only the high level function is `pub`.

**Comment by mitchellh** [comment]
Helpful! Its much faster to just pull out the bits you care about than to make a generalized "parse all features of `cpuid`" library. I've always found it easy to just do the cpuid stuff myself and do only the bit manipulation at the start of the program that I'll need for the rest of it.

**Comment by ominitay** [comment]
When implemented, this feature should probably allow selecting based not only on CPU features, but on the CPU family too. An example of the utility of this would be how AMD's processors implemented the BMI2 PDEP/PEXT instructions in microcode until Zen 3. This made the instructions incredibly slow, and therefore unfit for purpose. A project would likely want to treat this case as-if the instructions weren't available at all.

**Comment by yujiri8** [comment]
This is such an extremely obscure thing to want that I think it makes perfect sense to require implementing it manually via function pointers.

**Comment by nektro** [comment]
better yet,

```zig
pub fn someMathFunction(vec: Vector) Vector {
    if (builtin.cpu.arch == .x86_64) {
        if (comptime std.Target.x86.featureSetHas(builtin.cpu.features, .sse4_2)) {
            // ...
            // optimized for SSE 4.2
            //
            return;
        }
        if (comptime std.Target.x86.featureSetHas(builtin.cpu.features, .avx2)) {
            // ...
            // optimized for avx2
            //
            return;
        }
        // ...
        // no asm/intrinsics optimization
        //
    }
}

**Comment by andrewrk** [comment]
To be clear @mitchellh your use case is exactly what this issue is intended to solve.

Example would be swapping out memcpy on program startup to an implementation that takes full advantage of the CPU features detected at runtime.

The main challenge is doing so in a way that does not cause function calls to become virtual and thereby compromise perf which is the main purpose of this feature.

**Comment by sharpobject** [comment]
> This is such an extremely obscure thing to want that I think it makes perfect sense to require implementing it manually via function pointers.

The reason this wouldn't be great is that users of this feature probably want at least some monomorphization of callers of specialized functions. @andrewrk gave an example of memcpy, which should probably both be inlined into the caller and be a different implementation depending on the CPU.

**Comment by InKryption** [comment]
> I spitballed that one could try this by mutating a comptime global (does this even work?)

No, global mutable comptime memory is explicitly disallowed, as per #7396.

**Comment by ityonemo** [comment]
Am I incorrect that you could fairly easily manually implement multiversioning using comptime (anonymous) structs in an options parameter?

**Comment by 0xffox** [comment]
Despite this proposal dealing with versioning of functions based on CPU capabilities, I'd like to hear the opinion on extending it to versions of zig compiler itself.
```
fn myFunc(num: i32) i32 [zig: 0.11...0.13] {
    // blah
}

fn myFunc(num: i32) i32 [zig: 0.14] {
    // blah blah
}
```

Zig is rapidly changing language and it is difficult to support code that have compiled on previous version and do not on the new one. 
But I think zig could benefit from this versioning even pass 1.0 by embracing inevitability of evolution and prepare for it. 

Pros:
 + Explicit versioning 
 + Ease of keeping up with the language evolution
 + Updates of code base do not break existing builds/pipelines/container images etc


Cons:
 - Verbose
 -  Complicates error handling
 -  May result in messy codebase

**Comment by ljgermain** [comment]
Would compile-time function multi-versioning also be supported alongside runtime multi-versioning? For instance, if I compile something for the native target, the usual implication would be that I only intend to run it on my particular computer and CPU, in which case runtime CPU feature detection is just redundant overhead and doesn't make much sense.

Maybe a compiler flag should be provided that makes CPU feature detection happen at compile-time, which could be enabled-by-default when targeting the native CPU and disabled-by-default when targeting anything else? The set of features available would then just be detected from the host for the native case, and chosen from a pre-populated set for the case where a specific CPU architecture is targeted (this would represent what features are available as the lowest common denominator for said architecture).

The latter usually isn't desirable which is why I said runtime detection should be the default in that case (this ensures that the use case @mitchellh mentioned is well-supported), but I think it's good to have both. Whether or not runtime or compile-time detection is chosen should also be totally transparent to the application code (so there shouldn't be comptime if-statements checking features, for instance), which is easy to achieve with the original syntax presented by the creator of the issue or something similar.

Defaulting to compile-time selection of which function to use when targeting the host machine would also eliminate any potential overhead of virtual calls in that case which @andrewrk mentioned may be a concern (which I think is a neat optimization), although it would still exist in other cases assuming no further design. Thoughts?

**Comment by Cloudef** [comment]
Wrote this on discord
> Would be cool if zig had something like @specializeForTarget(.x86_64, .{ .avx512 = true, .. }) which you could specify multiple times and then it would generate (N) versions of the function (depending on target), and patch in start the most optimal version depending on cpu detection. There's some functions that are crazy fast in native compared to baseline.

**Comment by eira-fransham** [comment]
Maybe by analogy to `@inlineCall`:

```zig
if (cpu.hasFeatures(.foo, .bar)) {
  @specializeCall(fooFunc, [.foo, .bar], arg0, arg1, arg2)
}
```

Very PoC syntax but hopefully you get the point of making it a "call" builtin and letting the user handle when to use that specialised function (e.g maybe they want to have the dispatch be handled manually in order to benchmark different impls)
